### PR TITLE
Help for python exports

### LINKED ISSUE #140: Useful help text for Python exports
Right now there are nearly none.

This one works:
```python
import dataset as ds
help(ds.Variable.scalar)
```

### PR DESCRIPTION
**Pull request by nvaytet**
Added docstrings for most the the properties and functions for `Dimensions`, `Variable(Slice)` and `Dataset(Slice)`.

Examples:
```
help(Dataset.subset)
help(Variable.name)
```

Fixes #140

### CODE REVIEW

File: exports/dataset.cpp [hunk 1]
```diff
@@ -804,19 +849,25 @@ PYBIND11_MODULE(dataset, m) {
...
               const VariableSlice &other) {
-             const auto & [ dim, i ] = index;
+             const auto &[dim, i] = index;
              self(dim, i).assign(other);
            })
       .def("__setitem__", &detail::setVariableSlice)
       .def("__setitem__", &detail::setVariableSliceRange)
-      .def("copy", [](const VariableSlice &self) { return Variable(self); })
+      .def("copy", [](const VariableSlice &self) { return Variable(self); },
+           "Make a copy of a VariableSlice.")
```
Comment (OwenArnold) [inline exports/dataset.cpp]: "Copy variableslice as a variable"?

File: exports/dataset.cpp [hunk 2]
```diff
@@ -926,12 +979,20 @@ PYBIND11_MODULE(dataset, m) {
...
       .def_property_readonly(
-          "subset", [](DatasetSlice &self) { return SubsetHelper(self); })
+          "subset", [](DatasetSlice &self) { return SubsetHelper(self); },
+          "Used to extract a read-only subset of the DatasetSlice in two "
+          "different ways: \n - one can use just a string, e.g. "
+          "d.subset['sample'] to extract the sample and its variance, and all "
+          "the relevant coordinates.\n - one can also use a tag/string "
+          "combination for a more restrictive extraction, e.g. "
+          "d.subset[Data.Value, 'sample'] to only get the Value and associated "
+          "coordinates.")
```
Comment (OwenArnold) [inline exports/dataset.cpp]: Generally good. Is it worth stating what happens to attribute variables when the named subset option is used?

File: exports/dataset.cpp [hunk 3]
```diff
@@ -1084,12 +1147,19 @@ PYBIND11_MODULE(dataset, m) {
...
-      .def_property_readonly("subset",
-                             [](Dataset &self) { return SubsetHelper(self); })
+      .def_property_readonly(
+          "subset", [](Dataset &self) { return SubsetHelper(self); },
+          "Used to extract a read-only subset of the Dataset in two different "
+          "ways: \n - one can use just a string, e.g. d.subset['sample'] to "
+          "extract the sample and its variance, and all the relevant "
+          "coordinates.\n - one can also use a tag/string combination for a "
+          "more restrictive extraction, e.g. d.subset[Data.Value, 'sample'] to "
+          "only get the Value and associated coordinates.")
```
Comment (OwenArnold) [inline exports/dataset.cpp]: Same as above. Also, you get a slice back not a new `Dataset` don't know if you think that is worth mentioning?

File: exports/dataset.cpp [hunk 4]
```diff
@@ -1238,7 +1308,7 @@ PYBIND11_MODULE(dataset, m) {
       .def("__rmul__",
            [](const Dataset &self, double &other) { return self * other; },
            py::is_operator())
-      .def("merge", &Dataset::merge)
+      .def("merge", &Dataset::merge, "Merge two Datasets together.")
```
Comment (OwenArnold) [inline exports/dataset.cpp]: Can you be more specific about `merge` and `concatenate`. The meanings could be confused?

File: exports/dataset.cpp [hunk 5]
```diff
@@ -1255,16 +1325,21 @@ PYBIND11_MODULE(dataset, m) {
...
         py::overload_cast<const Dataset &, const Dataset &, const Dim>(
             &concatenate),
-        py::call_guard<py::gil_scoped_release>());
+        py::call_guard<py::gil_scoped_release>(),
+        "Returns a new dataset containing a concatenation of two Datasets "
+        "and their underlying Variables along a given Dimension.");
   m.def("rebin", py::overload_cast<const Dataset &, const Variable &>(&rebin),
-        py::call_guard<py::gil_scoped_release>());
+        py::call_guard<py::gil_scoped_release>(),
+        "Returns a new Dataset whose data is rebinned with new bin edges.");
```
Comment (OwenArnold) [inline exports/dataset.cpp]: minor point. "Rebinning" is probably not a ubiquitous term. Maybe "re-grid to new coordinate axis" is better?

File: exports/dataset.cpp (hunk 2 above)
Comment (nvaytet) [reply exports/dataset.cpp]: Can you tell me what happens to them So i can add it, thanks.

Comment (OwenArnold) [reply exports/dataset.cpp]: Looking at the code, I think they are not kept. Theres actually a comment about that here https://github.com/mantidproject/dataset/blob/master/src/dataset.h#L332

**Pull request merged**